storagemarket/deal_execution.go

```diff
@@ -685,66 +687,88 @@ func (p *Provider) registerShardSync(ctx context.Context, pc cid.Cid) error {
 
 // fireSealingUpdateEvents periodically checks the sealing status of the deal
 // and fires events for each change
-func (p *Provider) fireSealingUpdateEvents(dh *dealHandler, dealUuid uuid.UUID, sectorNum abi.SectorNumber) {
-	var lastSealingState lapi.SectorState
-	checkStatus := func(force bool) lapi.SectorState {
+func (p *Provider) fireSealingUpdateEvents(dh *dealHandler, dealUuid uuid.UUID, sectorNum abi.SectorNumber) *dealMakingError {
+	var sectorInfo lapi.SectorInfo
+	var deal *types.ProviderDealState
+	checkStatus := func(force bool) lapi.SectorInfo {
 		// To avoid overloading the sealing service, only get the sector status
 		// if there's at least one subscriber to the event that will be published
 		if !force && !dh.hasActiveSubscribers() {
-			return ""
+			return lapi.SectorInfo{}
 		}
 
 		// Get the sector status
 		si, err := p.sps.SectorsStatus(p.ctx, sectorNum, false)
-		if err == nil && si.State != lastSealingState {
-			lastSealingState = si.State
+		if err == nil && si.State != sectorInfo.State {
+			sectorInfo.State = si.State
 
 			// Sector status has changed, fire an update event
-			deal, err := p.dealsDB.ByID(p.ctx, dealUuid)
+			deal, err = p.dealsDB.ByID(p.ctx, dealUuid)
 			if err != nil {
 				log.Errorf("getting deal %s with sealing update: %w", dealUuid, err)
-				return si.State
+				return si
 			}
 
 			p.dealLogger.Infow(dealUuid, "current sealing state", "state", si.State)
 			p.fireEventDealUpdate(dh.Publisher, deal)
 		}
-		return si.State
+		return si
+	}
+
+	retErr := &dealMakingError{
+		retry: types.DealRetryFatal,
+		error: ErrDealNotSealed,
 	}
 
 	// Check status immediately
-	state := checkStatus(true)
-	if isFinalSealingState(state) {
-		return
+	info := checkStatus(true)
+	if IsFinalSealingState(info.State) {
+		if len(info.Deals) > 0 {
+			for _, d := range info.Deals {
+				if d == deal.ChainDealID {
+					return nil
+				}
+			}
+			return retErr
+		}
+		return retErr
```

Comment:
This code is repeated several times in this PR. I'd suggest refactoring the code into a function that can be called from all those places.
Or maybe IsFinalSealingState should return `(bool, error)`? I guess try it out and see which makes the code simpler.

Reply:
I have removed the initial if condition and shorten this code to 3 lines. Inserting this check in IsFinalSealingState() will make things more complicated** as we will need to supply the ProvideDealState to the function as well. Check the new code once and let me know if it is cleaner. Otherwise, I will make the changes.

**IsFinalSealingState() is also being used outside the deal_execution now